src/patch_parse.c

```diff
@@ -0,0 +1,1025 @@
+#include "git2/patch.h"
+#include "patch.h"
+#include "path.h"
+
+#define parse_err(...) \
+	( giterr_set(GITERR_PATCH, __VA_ARGS__), -1 )
+
+typedef struct {
+	git_patch base;
+
+	git_patch_options opts;
+
+	/* the patch contents, which lines will point into. */
+	/* TODO: allow us to point somewhere refcounted. */
+	char *content;
+
+	/* the paths from the `diff --git` header, these will be used if this is not
+	 * a rename (and rename paths are specified) or if no `+++`/`---` line specify
+	 * the paths.
+	 */
+	char *header_old_path, *header_new_path;
+
+	/* renamed paths are precise and are not prefixed */
+	char *rename_old_path, *rename_new_path;
+
+	/* the paths given in `---` and `+++` lines */
+	char *old_path, *new_path;
+
+	/* the prefixes from the old/new paths */
+	char *old_prefix, *new_prefix;
+} git_patch_parsed;
+
+typedef struct {
+	const char *content;
+	size_t content_len;
+
+	const char *line;
+	size_t line_len;
+	size_t line_num;
+
+	size_t remain;
+} patch_parse_ctx;
+
+
+static void parse_advance_line(patch_parse_ctx *ctx)
+{
+	ctx->line += ctx->line_len;
+	ctx->remain -= ctx->line_len;
+	ctx->line_len = git__linenlen(ctx->line, ctx->remain);
+	ctx->line_num++;
+}
+
+static void parse_advance_chars(patch_parse_ctx *ctx, size_t char_cnt)
+{
+	ctx->line += char_cnt;
+	ctx->remain -= char_cnt;
+	ctx->line_len -= char_cnt;
+}
+
+static int parse_advance_expected(
+	patch_parse_ctx *ctx,
+	const char *expected,
+	size_t expected_len)
```

Comment:
It'd be nice to have a version of this function which we can use for static strings so we don't have to count by hand in order to have the value be known at compile time, like we do with `git_buf_PUTS()`.